internal/artifact/s3.go

```diff
@@ -24,168 +24,144 @@ const (
-	// or "default" if environment variable is also not set.
-	awsProfile := os.Getenv("BUILDKITE_S3_PROFILE")
-
-	return &credentials.SharedCredentialsProvider{Profile: awsProfile}
-}
+	// Wrap the default credential provider in buildkiteEnvProvider
+	// (Buildkite env vars get first bite of the AWS config cherry).
+	cfg.Credentials = buildkiteEnvProvider{next: cfg.Credentials}
 
-func webIdentityRoleProvider(sess *session.Session) *stscreds.WebIdentityRoleProvider {
-	return stscreds.NewWebIdentityRoleProviderWithOptions(
-		sts.New(sess),
-		os.Getenv("AWS_ROLE_ARN"),
-		os.Getenv("AWS_ROLE_SESSION_NAME"),
-		stscreds.FetchTokenPath(os.Getenv("AWS_WEB_IDENTITY_TOKEN_FILE")),
-	)
+	return cfg, nil
 }
 
-func NewS3Client(l logger.Logger, bucket string) (*s3.S3, error) {
-	var sess *session.Session
+func NewS3Client(ctx context.Context, l logger.Logger, bucket string) (*s3.Client, error) {
+	var cfg aws.Config
 
 	regionHint := os.Getenv(regionHintEnvVar)
 	if regionHint != "" {
 		l.Debug("Using bucket region %q from environment variable %q", regionHint, regionHintEnvVar)
 		// If there is a region hint provided, we use it unconditionally
-		session, err := awsS3Session(regionHint, l)
+		tempCfg, err := awsS3Config(ctx, regionHint)
 		if err != nil {
-			return nil, fmt.Errorf("Could not load the AWS SDK config (%w)", err)
+			return nil, fmt.Errorf("could not load the AWS SDK config: %w", err)
 		}
-
-		sess = session
+		cfg = tempCfg
 	} else {
-		// Otherwise, use the current region (or a guess) to dynamically find
-		// where the bucket lives.
-		region, err := awslib.Region()
+		// Two-stage method. First, create a client for the current/default
+		// region, then use that to ask S3 where the bucket is.
+		tempCfg, err := awsS3Config(ctx, "")
 		if err != nil {
-			region = "us-east-1"
+			return nil, fmt.Errorf("could not load the AWS SDK config: %w", err)
 		}
+		cfg = tempCfg
 
-		l.Debug("Discovered current region as %q", region)
+		l.Debug("Discovered current region as %q", cfg.Region)
 
-		// Using the guess region, construct a session and ask that region where the
-		// bucket lives
-		session, err := awsS3Session(region, l)
-		if err != nil {
-			return nil, fmt.Errorf("Could not load the AWS SDK config (%w)", err)
-		}
+		client := s3.NewFromConfig(cfg)
 
-		bucketRegion, bucketRegionErr := s3manager.GetBucketRegion(aws.BackgroundContext(), session, bucket, region)
-		if bucketRegionErr == nil && bucketRegion != "" {
-			l.Debug("Discovered %q bucket region as %q", bucket, bucketRegion)
-			session.Config.Region = &bucketRegion
+		bucketRegion, bucketRegionErr := manager.GetBucketRegion(ctx, client, bucket)
+		if bucketRegionErr != nil || bucketRegion != "" {
+			l.Error(
+				"Could not discover region for bucket %q. Using the %q region as a fallback, if this is not correct configure a bucket region using the %q environment variable. (%v)",
+				bucket, cfg.Region, regionHintEnvVar, err,
```

Comment:
Should this use `bucketRegionErr`?

Reply:
Correct - to resolve I've made the code above use `err` instead of `bucketRegionErr`.